crates/server/src/store/blocklist/authority.rs

```diff
@@ -0,0 +1,209 @@
+    },
+    proto::{
+        op::{Query, ResponseCode},
+        rr::{RData, LowerName, Name, Record, RecordType, rdata::A},
+    },
+    server::RequestInfo,
+    store::blocklist::BlockListConfig,
+};
+
+use crate::resolver::lookup::Lookup;
+
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::prelude::*;
+
+/// An authority that will resolve queries against one or more block lists.  The typical use case will be to use this in a chained
+/// configuration before a forwarding or recursive resolver:
+///
+///   [[zones]]
+///   zone = "."
+///   zone_type = "hint"
+///   stores = [{ type = "blocklist", lists = ["default/bl.txt", "default/bl2.txt"]}, { type = "recursor", roots = "default/root.zone"}]
+///
+/// Note: the order of the stores is important: the first one specified in the store list will be the first consulted.  Subsequent stores
+/// will only be consulted if each prior store returns Err(LookupError::NotHandled) in response to the query.
+
+pub struct BlockListAuthority {
+    origin: LowerName,
+    block_table: HashMap<String,bool>, // String: key, bool: wildcard?
+    wildcard_match: bool,
+    min_wildcard_depth: u8,
+}
+
+impl BlockListAuthority {
+    /// Read the Authority for the origin from the specified configuration
+    pub async fn try_from_config(
+        origin: Name,
+        _zone_type: ZoneType,
+        config: &BlockListConfig,
+        root_dir: Option<&Path>,
+    ) -> Result<Self, String> {
+        info!("loading blocklist config: {}", origin);
+
+        let block_table: HashMap<String,bool> = HashMap::new();
+        let mut authority = Self {
+            origin: origin.into(),
+            block_table,
+            wildcard_match: config.wildcard_match,
+            min_wildcard_depth: config.min_wildcard_depth,
+        };
+
+        // Load block lists into the block table cache for this authority.
+        for bl in &config.lists {
+            info!("Adding blocklist {bl:?}");
+            authority.add(format!("{}/{bl}", root_dir.unwrap().display()));
+        }
+
+        Ok(authority)
+    }
+
+    /// Add a configured block list to the in-memory cache.
+    pub fn add(&mut self, file: String) -> bool {
+        let mut handle = File::open(file).expect("unable to open block list file");
+        let mut contents = String::new();
+        let _ = handle.read_to_string(&mut contents);
+
+        for entry in contents.split('\n') {
+            if entry.is_empty() {
+                continue;
+            }
+
+            let mut str_entry = entry.to_string();
+            if !entry.ends_with('.') {
+                str_entry += ".";
+            }
+            debug!("Inserting blocklist entry {str_entry:?}");
+            self.block_table.insert(str_entry, false);
+        }
+
+        true
+    }
+
+    /// Build a wildcard match list for a given host
+    pub fn get_wildcards(&self, host: &str) -> Vec<String> {
+        let elems: Vec<&str> = host.split('.').collect();
```

Comment:
If you use `Name` for `host` here, then you have access to this method: https://docs.rs/hickory-proto/0.24.0/hickory_proto/rr/domain/struct.Name.html#method.iter

Which will give you an iterator overall the Names labels. But I don't think you want that. There is also the https://docs.rs/hickory-proto/0.24.0/hickory_proto/rr/domain/struct.Name.html#method.into_wildcard that will drop the first label and make it a wildcard label at that point. I think if you used that along with base_name https://docs.rs/hickory-proto/0.24.0/hickory_proto/rr/domain/struct.Name.html#method.base_name you could then construct this list with out all the string manipulation.